src/instructions/port.rs

```diff
@@ -95,84 +95,72 @@ impl PortWrite for u32 {
     }
 }
 
-/// A read only I/O port.
-pub struct PortReadOnly<T> {
-    port: u16,
-    phantom: PhantomData<T>,
-}
+/// A marker trait for access types which allow reading port values.
+pub trait PortReadAccess {}
 
-/// A write only I/O port.
-pub struct PortWriteOnly<T> {
-    port: u16,
-    phantom: PhantomData<T>,
-}
+/// A marker trait for access types which allow writing port values.
+pub trait PortWriteAccess {}
 
-/// An I/O port.
-pub struct Port<T> {
-    port: u16,
-    phantom: PhantomData<T>,
+/// An access marker type indicating that a port is only allowed to read values.
+#[derive(Debug)]
+pub struct ReadOnlyAccess {
```

Comment:
Could we write these Access structs as `pub struct ReadOnlyAccess(());`? It's a little shorter.